### PR TITLE
Add ERC: Smart Contract Delegation

### PR DESCRIPTION
**Pull request by rojotek**
Adds a new EIP that specifies a delegation interface for smart contract accounts.

This proposal defines interfaces for smart contracts to delegate capabilities to other smart contracts,
smart contract accounts, or Externally Owned Accounts (EOAs). The delegating contract MUST
implement ERC-1271 along with the `IDelegator` interface. This framework empowers a
delegating contract with the ability to delegate any actions it has the authority to perform,
thereby enabling more flexible and scalable contract interactions. This standard outlines the
minimal interface necessary to facilitate such delegation.

Additionally, this proposal is compatible with ERC-4337, although its implementation does not
necessitate ERC-4337.

### CODE REVIEW

File: ERCS/eip-draft-delegation-framework.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
+---
```
Comment (abcoathup) [inline ERCS/eip-draft-delegation-framework.md:1]:
```suggestion
---
eip: 7710
```
Assigning next sequential EIP/ERC/RIP number.

Please also update the filename.

**Comment by danfinlay** [comment]
Looks like this should be good to merge?

**Review by danfinlay — APPROVED** [review]

**Comment by danfinlay** [comment]
cc @axic could we get a merge here?

**Comment by McOso** [comment]
@SamWilsn anything else needed to merge?

File: ERCS/erc-7710.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
+---
+eip: 7710
+title: Smart Contract Delegation Interfaces
+description: Minimal interfaces to enable smart contracts to consistently delegate capabilities to other contracts or EOAs.
```
Comment (SamWilsn) [inline ERCS/erc-7710.md:4]:
```suggestion
description: Interfaces for consistently delegating capabilities to other contracts or EOAs.
```

Small suggestion here to remove some fluff.

File: ERCS/erc-7710.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
+---
+eip: 7710
+title: Smart Contract Delegation Interfaces
```
Comment (SamWilsn) [inline ERCS/erc-7710.md:3]:
```suggestion
title: Smart Contract Delegation
```

File: ERCS/erc-7710.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+### Overview
+
+#### Redeeming a Delegation
+
+When a delegate wishes to redeem a delegation, they call the `redeemDelegation` function on the Delegation Manager and
+pass in the Action they want to execute and the proof of authority (ie delegation) which they are executing on behalf
+of. The Delegation Manager then verifies the delegation's validity and, if valid, calls the privileged function on the
+Delegator which executes the specified capability on behalf of the Delegator.
+
+```mermaid
```
Comment (SamWilsn) [inline ERCS/erc-7710.md:70]: Our renderer doesn't support mermaid. Please replace this with an image, preferably an SVG.

File: ERCS/erc-7710.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+eip: 7710
+title: Smart Contract Delegation Interfaces
+description: Minimal interfaces to enable smart contracts to consistently delegate capabilities to other contracts or EOAs.
+author: Ryan McPeck (@McOso), Dan Finlay (@DanFinlay), Rob Dawson (@rojotek), Derek Chiang (@derekchiang)
+discussions-to: https://ethereum-magicians.org/t/towards-more-conversational-wallet-connections-a-proposal-for-the-redeemdelegation-interface/16690
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-05-20
+requires: 1271
```
Comment (SamWilsn) [inline ERCS/erc-7710.md:11]:
```suggestion
requires: 1271, 7579
```

Line 92 strongly implies that reading 7579 is required to understand the meaning of `mode` and `executionCalldata`.

File: ERCS/erc-7710.md [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+#### `ERC7710Manager.sol`
+
+The Delegation Manager MUST implement the `redeemDelegation` which will be responsible for validating the delegation
+being invoked, and will then call the delegator to execute the action.
+
+The bytes `_authority` passed in as a parameter to the `redeemDelegation` function contains the authority to execute a
+specific action on behalf of the delegating contract.
+
+The `Action` struct has two fields: `mode` and `executionCalldata`, which are defined precisely as in ERC-7579 (under the "Execution Behavior" section).  Briefly, `mode` encodes the "behavior" of the execution, which could be a single call, a batch call, a delegatecall, and others.  `executionCallData` encodes the data of the execution, which typically includes at least a `target`, a `value`, and a `to` address.
```
Comment (SamWilsn) [inline ERCS/erc-7710.md:92]:
```suggestion
The `Action` struct has two fields: `mode` and `executionCalldata`, which are defined precisely as in [ERC-7579](./eip-7579.md) (under the "Execution Behavior" section).  Briefly, `mode` encodes the "behavior" of the execution, which could be a single call, a batch call, a delegatecall, and others.  `executionCallData` encodes the data of the execution, which typically includes at least a `target`, a `value`, and a `to` address.
```

File: ERCS/erc-7710.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+executions.
+
+## Security Considerations
+
+The introduction of customizable authorization terms requires careful consideration of how authorization data is
+structured and interpreted. Potential security risks include the misinterpretation of authorization terms and
+unauthorized actions being taken if the interface is not properly implemented. It is recommended that applications
+implementing this interface undergo thorough security audits to ensure that authorization terms are handled securely.
+
+Needs discussion.
```
Comment (SamWilsn) [inline ERCS/erc-7710.md:173]:
```suggestion
Needs discussion. <!-- TODO -->
```

If you use an HTML-style comment, the linter will make sure you remove it before moving to Review.

File: ERCS/erc-7710.md (hunk 4 above)
Comment (rojotek) [reply ERCS/erc-7710.md]: have updated and added in an SVG in the assets folder for this.

File: ERCS/erc-7710.md (hunk 5 above)
Comment (rojotek) [reply ERCS/erc-7710.md]: Have added the require into the latest version.

File: ERCS/erc-7710.md (hunk 7 above)
Comment (rojotek) [reply ERCS/erc-7710.md]: comment added.

**Review by SamWilsn — APPROVED** [review]

**Pull request merged**